### ISSUE TITLE
ITK inlines an SSE2 function wheh SSE2 isn't enabled in the client code

### ISSUE THREAD

**Issue by yurivict** [InsightSoftwareConsortium/ITK#2011]
The compilation of the ```vv``` project (https://github.com/open-vv/vv) fails in the ITK code, see log http://beefy5.nyi.freebsd.org/data/121i386-default/549547/logs/vv-1.4.0.383.log

There is no ```-sse2```  flag but ```USE_SSE2_32IMPL``` is set in the ITK headers and the compiler errors out,

FreeBSD 12
clang 10
vv-1.4.0-383-g7a783a8b
InsightToolkit-5.0.1_6

**Comment by dzenanz** [comment]
This might be due to imperfect support of FreeBSD by CMake:
https://github.com/InsightSoftwareConsortium/ITK/blob/344e9bc58e69ed921f540c67d87dbb66fee4c466/Modules/Core/Common/src/itkConfigure.h.in#L155-L157
@bradking what do you think?

**Comment by dzenanz** [comment]
@yurivict how do you disable SSE2, and why?

**Comment by bradking** [comment]
@dzenanz I don't see what your quoted code has to do with FreeBSD+CMake.

**Comment by dzenanz** [comment]
CMake might be misdetecting `HAVE_EMMINTRIN_H`.

**Comment by bradking** [comment]
CMake doesn't detect `HAVE_EMMINTRIN_H`.  ITK's CMake code does that.

**Comment by yurivict** [comment]
> @yurivict how do you disable SSE2, and why?

I didn't disable SSE2. I am only letting ```vv``` build and it apparently doesn't enable SSSE2 by not adding ```-msse2```, so it remains disabled by default, but ITK still assumes thatit is enabled.

**Comment by yurivict** [comment]
> CMake doesn't detect HAVE_EMMINTRIN_H. ITK's CMake code does that.

This should be reported to cmake then?

**Comment by bradking** [comment]
> This should be reported to cmake then?

No.  The code checking for `HAVE_EMMINTRIN_H` is in ITK's `Modules/Core/Common/CMake/CheckSupportForSSERounding.cmake` module.

**Comment by dzenanz** [comment]
I think that ITK assumes that it will be built on hardware less than 10 years old, so it should have SSE2 instructions. Was this being compiled for i386 architecture model? I think that is what prevents use of SSE2 instructions.

**Comment by yurivict** [comment]
Yes, it the log linked above is for i386. The hardware isn't old, it builds for i386 regardless.

The problem seems to be that clang doesn't enable SSE2 by default w/out ```-msse2```, and it shouldn't.

**Comment by dzenanz** [comment]
I guess the conditions here should be satisfied:
https://github.com/InsightSoftwareConsortium/ITK/blob/v5.0.1/CMake/ITKSetStandardCompilerFlags.cmake#L304-L307

**Comment by issakomi** [comment]
FYI, _-msse2_ should be enabled only for x86-32 compiler, for x86-64 it is default (already several years), x86-32 builds are rare now, probably it was forgotten somehow. I run into the issues as i tried to build code for x86-32 with Sony's vectormath not so long ago

https://gcc.gnu.org/onlinedocs/gcc-9.2.0/gcc/x86-Options.html
`For the x86-32 compiler, you must use -march=cpu-type, -msse or -msse2 switches to enable SSE extensions and make this option effective. For the x86-64 compiler, these extensions are enabled by default. `

i guess, the same for clang

**Comment by yurivict** [comment]
I think you are right.

**Comment by issakomi** [comment]
FYI. On Linux, with ITK installed from Debian package (x86-64), where is
_/usr/lib/cmake/ITK-4.12/ITKConfig.cmake_ file with empty flags
```
# The C and C++ flags added by ITK to the cmake-configured flags.
set(ITK_REQUIRED_C_FLAGS "")
set(ITK_REQUIRED_CXX_FLAGS "")
set(ITK_REQUIRED_LINK_FLAGS "")
```

My local ITK-5.2 build (x86-64) has
```
# The C and C++ flags added by ITK to the cmake-configured flags.
set(ITK_REQUIRED_C_FLAGS "")
set(ITK_REQUIRED_CXX_FLAGS " -msse2")
set(ITK_REQUIRED_LINK_FLAGS "")
```
i didn't set any related option manually

Seems it is sometimes enabled, sometimes not, depending on ITK version, platform, etc.

**Comment by dzenanz** [comment]
> I think you are right.

Should this issue be closed?

**Comment by yurivict** [comment]
I've confirmed that this is correct that SSE2 is enabled by default on amd64 and disabled by default on i386. SSE2 is considered a part of the basic instruction set on amd64 but not on i386.

Therefore, since ITK depends on SSE2, the correct resolution of this problem is that you should add the flag ```-msse2``` to all ```CFLAGS``` and ```CXXFLAGS```.

**Comment by dzenanz** [comment]
@hjmjohnson I remember you were dealing with those flags. Also we seem to be taking the [conditions from VNL](https://github.com/InsightSoftwareConsortium/ITK/blob/v5.0.1/CMake/ITKSetStandardCompilerFlags.cmake#L304-L307). What is the right place to add this?

**Comment by dzenanz** [comment]
https://github.com/vxl/vxl/pull/796 is a PR regarding this.

**Comment by hjmjohnson** [comment]
This particular issue is not me.  VXL does some auto-setting of sse2 flags from code in 2008.  That code seems to be wrong.

@thewtex added the code you reference in 2015, which seems like perhaps it was a work around for VXL build failures.

I think that we should *NOT* set -msse2 based on which compiler is found.

**Comment by dzenanz** [comment]
Has this issue been fixed?

**Comment by hjmjohnson** [comment]
Not fixed or addressed by me.

**Comment by thewtex** [comment]
@yurivict have you tried with ITK Git `master`?

**Comment by thewtex** [comment]
Related: https://github.com/simd-everywhere/simde

**Comment by DominionSoftware** [comment]
I think I'm getting bitten by this compiled 5.1.2 on windows. 

\itk512.ins\include\ITK-5.1\vnl\vnl_sse.h(217,1): error C2059: syntax error: 'const' (compiling source file

**Comment by DominionSoftware** [comment]
I have VNL_CONFIG_ENABLE_SSE2 off and ...._ROUNDING off

**Comment by hjmjohnson** [comment]
Closing as fixed.

Commit `6c800e4640` ("BUG: ITK_COMPILER_SUPPORTS_SSE2_64 incorrectly set for 32-bit x86", 2022-11-28) gates the SSE2 macros on whether the target compiler is actually generating SSE2 instructions. Current `Modules/Core/Common/src/itkConfigure.h.in` lines ~182-189:

```c
#if defined(__x86_64__) || defined(__amd64__) || (defined(_M_AMD64) && !defined(_M_ARM64EC))
#  define ITK_COMPILER_SUPPORTS_SSE2_32
#  define ITK_COMPILER_SUPPORTS_SSE2_64
#else
#  if defined(__SSE2__) || (defined(_M_IX86_FP) && (_M_IX86_FP >= 2))
#    define ITK_COMPILER_SUPPORTS_SSE2_32
#  endif
#endif
```

So 32-bit x86 client code without `-msse2` no longer gets `ITK_COMPILER_SUPPORTS_SSE2_32` defined, and will not pull in SSE2 intrinsics through the installed configure header. Same fix resolves #2529.

Triaged as part of the 2026-04-10 open-issue relevance review.
